### DISCUSSION TITLE
KEYB.COM / CHCP.COM - keyboard combination to switch codepages, synchronize layout with host

### DISCUSSION THREAD

**Discussion by Torinde** [dosbox-staging/dosbox-staging#3239] [Ideas]
1. When using a non-Latin script codepage (e.g. Cyrillic) the box drawing characters of most installers and programs are missing/rearranged (because they're hardcoded for 437). 

It would be great if KEYB or Staging itself supports a key-combination to change between 437 and whatever codepage is active - for example 3021 moves around the box drawing characters and back in the day we used TSR programs allowing `LeftShift `+ `RightShift` to change between the codepages on-the-fly (when you need to see something like a game installer menu, which is written in 437, but soon after you'll change back to 3021 in order to read/write in Cyrillic). I mentioned it before [here](https://github.com/dosbox-staging/dosbox-staging/pull/3164#issuecomment-1837540852). 

2. `Alt`+`Shift` (current way to change scripts)- either sync with host (changing on host changes also in Staging, changing in Staging changes also on host) or decouple fully from host (changes on host never affect Staging, changes in Staging never affect host). Currently if I press in Staging Alt+RightShift that changes to Cyrillic both in Staging and on host (coincidence, because both Staging KEYB and Windows use the same key combo to switch?) - nice, so I can type with the same alphabet in both, but then while I'm on the host and I switch back to Latin (Alt+RightShift - Windows uses the same combo to revert - unlike KEYB which needs Alt+LeftShift) Staging doesn't follow and in general it becomes annoying if you go between host/Staging and type in both - soon they get out of sync with each other and you think/expect your layout is Latin, but it's not, etc.

I don't see an obvious "best" solution here, so just sharing to see what you think and most importantly is that something you want to improve.

Same topic: https://github.com/dosbox-staging/dosbox-staging/issues/3128#issuecomment-1825396175, joncampbell123/dosbox-x/issues/3634. https://github.com/dosbox-staging/dosbox-staging/issues/3523#issuecomment-2039198481, Related #3233, https://github.com/dosbox-staging/dosbox-staging/issues/3498#issuecomment-1991000707, https://github.com/dosbox-staging/dosbox-staging/pull/3951#issuecomment-2392294903, https://github.com/dosbox-staging/dosbox-staging/pull/4678#issuecomment-3762971014

To illustrate why LeftShift+RightShift combo is needed for user to switch to/from 437 - see how EDIT looks with 3021:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/763445af-790f-465f-ac46-5f02bd9267f1)

While how it should look:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/aa17357e-cdde-473f-b1ae-8aa430c21b6c)

And side-by-side - 437 vs 3021, including both Latin and Cyrillic text:
![image](https://github.com/user-attachments/assets/39acec86-392d-4d7a-9cd6-e1c985414644)

Both views are useful and user needs ability to change between them at will / on-the-fly.

**Comment by NicknineTheEagle** [comment]
> When using a non-Latin script codepage (e.g. Cyrillic) the box drawing characters of most installers and programs are missing/rearranged (because they're hardcoded for 437).

CP 866 actually keeps all the pseudographic characters (unlike CP 855 which is why it never took off in Russia).
But I know that some European code pages replace them (e.g. CP 850).

**Comment by Pellzi** [comment]
To use more Codepages in Dos, it´s required, that this is in the Config.sys and autoexec.bat file written, that you want do this.

It s a long time ago, so i can´t remember, why these lines looking like this.
But chcp 850, or chcp 427 switch the used codepage.

[Config.zip](https://github.com/user-attachments/files/17688971/Config.zip)

**Reply by Torinde** [reply]
Interesting, but from what I see you load two code pages, but have to _choose one_ of them via DOS commands/programs (CHCP and maybe Graftabl.com) _at the prompt_. What I need is a way to _switch on-the-fly_ (while a program is running).

**Reply by Pellzi** [reply]
All i know about DOS is, before you can use a Codepage site inside a *.cpi File, you have to load this site into memory.
These Lines in the config.zip text File for Config.sys and autoexec.bat makes this happens.
If this is not the case, DOS don´t know, what you want from it, and how the character should look like.

Probably, when the codepages, you use, are not available in a standard *.cpi File, you have to produce your own *.cpi file, which includes after this exercise, what you want. But even then, these lines at the start of DOS are required.
To do this, cpied.exe is the best DOS program i know, to do this.
https://help.fdos.org/en/hhstndrd/util/cpied.htm   Because, on this Side, there are no download link, i separated this program out of the Freeoos Distribution, that you have not such a great work to do it by yourself.

[CPIED.zip](https://github.com/user-attachments/files/17692597/CPIED.zip)

The only help, i can give you, if this also not helps, you should try maybee a another Editor, which can handle this problem.
It´s called Bloček. His description is:
Bloček is a text editor for DOS with unicode support. It means that unlike to other DOS editors, you are not limited to your current code page. It allows you to edit multilanguage documents, translate texts from one language to another, write modern web pages and so on.

And the Addon for it, Kašmár
Kašmár is tool for developers. It is a unicode font editor which allows you to create or modify bitmap unicode fonts used in my programs and libraries. It has also functions for import and export from/to VGA font files (which are used by many legacy utilities) so it can be considered also as a VGA font editor.

http://www.laaca.borec.cz/blocek/

**Reply by Torinde** [reply]
Thank you for trying to help!

Even with the current tools in Staging I can load either codepage and switch between them (_from the prompt_), that works OK. My problem is that I need to switch between them _from inside a running program_. Do you know a key combo (like Alt+Shift) that results in a codepage change when using the standard DOS tools?

**Reply by Pellzi** [reply]
Simply answer, NO.

I can only imagine a solution, which is not real, and its only a idea for a possible doing of this behaviour of DOS.
To explain this, it´s only a guess, like it could be maybe done. But don´t condemn me, if this, what i write is not true.
Have fun with this text. And interpret them not like a CPU, instead like a thinking human. :-)

[64_Ch_Cp.zip](https://github.com/user-attachments/files/17705510/64_Ch_Cp.zip)
